File: WuchangLiveWallpaper/app/src/main/java/com/halot455/wuchangwallpaper/HeroRenderer.java
package com.halot455.wuchangwallpaper;

import android.content.Context;
import android.graphics.*;
import android.util.Base64;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public final class HeroRenderer {
    private final Context context;
    private final Map<String, Bitmap> cache = new HashMap<>();
    private final Paint p = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.FILTER_BITMAP_FLAG);
    private final Paint glow = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint ribbon = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Random random = new Random(455);
    private final float[] px = new float[42];
    private final float[] py = new float[42];
    private final float[] ps = new float[42];
    private final float[] pv = new float[42];

    public HeroRenderer(Context context) {
        this.context = context.getApplicationContext();
        for (int i = 0; i < px.length; i++) {
            px[i] = random.nextFloat();
            py[i] = random.nextFloat();
            ps[i] = 1.1f + random.nextFloat() * 3.4f;
            pv[i] = 0.016f + random.nextFloat() * 0.032f;
        }
    }

    private Bitmap load(String key) {
        Bitmap b = cache.get(key);
        if (b != null) return b;
        try (InputStream in = context.getAssets().open(key + ".b64")) {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            byte[] buf = new byte[8192];
            int n;
            while ((n = in.read(buf)) > 0) out.write(buf, 0, n);
            String s = out.toString(StandardCharsets.UTF_8.name())
                    .replace("\n", "").replace("\r", "").trim();
            byte[] data = Base64.decode(s, Base64.DEFAULT);
            b = BitmapFactory.decodeByteArray(data, 0, data.length);
            if (b != null) cache.put(key, b);
            return b;
        } catch (Exception ignored) {
            return null;
        }
    }

    private static float clamp(float v, float lo, float hi) {
        return Math.max(lo, Math.min(hi, v));
    }

    public void draw(Canvas c, int w, int h, float angle, long now, String character, float pageOffset) {
        if (w <= 0 || h <= 0) return;
        boolean white = !"black".equals(character);
        float t = now / 1000f;
        float turn = clamp(angle / 35f, -1f, 1f);

        p.setShader(new LinearGradient(
                0, 0, 0, h,
                white ? Color.rgb(18, 26, 43) : Color.rgb(18, 15, 30),
                Color.rgb(4, 7, 13), Shader.TileMode.CLAMP));
        c.drawRect(0, 0, w, h, p);
        p.setShader(null);

        float moonR = Math.min(w, h) * 0.13f;
        float moonX = w * (0.76f - pageOffset * 0.05f - turn * 0.012f);
        float moonY = h * 0.15f;
        glow.setShader(new RadialGradient(
                moonX, moonY, moonR * 1.55f,
                new int[]{0xEEFFFFFF, 0x99DDE7FF, 0x002E3C65},
                null, Shader.TileMode.CLAMP));
        c.drawCircle(moonX, moonY, moonR * 1.55f, glow);
        glow.setShader(null);

        glow.setShader(new RadialGradient(
                w * 0.5f, h * 0.83f, w * 0.39f,
                new int[]{white ? 0x397E9EFF : 0x3D6B2CB5, 0x00000000},
                null, Shader.TileMode.CLAMP));
        c.drawOval(w * .10f, h * .70f, w * .90f, h * .95f, glow);
        glow.setShader(null);

        Bitmap b = load(character + "_front");
        if (b != null) {
            float breathe = 1f + 0.0045f * (float)Math.sin(t * 1.55f);
            float targetH = h * 0.86f;
            float scale = targetH / b.getHeight();
            float widthCompress = 1f - Math.abs(turn) * 0.095f;
            float targetW = b.getWidth() * scale * widthCompress;
            float left = (w - targetW) * 0.5f + turn * w * 0.018f;
            float top = h * 0.075f + (float)Math.sin(t * 1.25f) * h * 0.0025f;
            Rect src = new Rect(0, 0, b.getWidth(), b.getHeight());
            RectF dst = new RectF(left, top, left + targetW, top + targetH);

            c.save();
            c.translate(turn * w * 0.007f, 0f);
            c.skew(-turn * 0.018f, 0f);
            c.scale(1f, breathe, w * .5f, top + targetH * .55f);
            c.drawBitmap(b, src, dst, p);
            c.restore();

            float split = top + targetH * .43f;
            float sway = (float)Math.sin(t * 1.08f + angle * .013f);
            c.save();
            c.clipRect(0, split, w, h);
            c.translate((sway + turn * .35f) * w * .009f, 0);
            c.skew((sway * .009f) - turn * .006f, 0);
            p.setAlpha(82);
            c.drawBitmap(b, src, dst, p);
            p.setAlpha(255);
            c.restore();

            c.save();
            c.clipRect(0, top, w, top + targetH * .36f);
            c.translate((float)Math.sin(t * 1.35f) * w * .0032f - turn * w * .002f, 0);
            c.skew(-turn * .004f, 0);
            p.setAlpha(68);
            c.drawBitmap(b, src, dst, p);
            p.setAlpha(255);
            c.restore();
        }

        ribbon.setStyle(Paint.Style.STROKE);
        ribbon.setStrokeWidth(Math.max(2f, w * .0023f));
        ribbon.setColor(white ? 0x88EAF1FF : 0xAA9A55FF);
        for (int k = 0; k < 3; k++) {
            float phase = t * (0.42f + k * .05f) + k * 1.7f;
            float yy = h * (.55f + k * .09f);
            Path path = new Path();
            path.moveTo(w * .08f, yy + (float)Math.sin(phase) * 18f);
            path.cubicTo(
                    w * .26f, yy - 48f,
                    w * .64f, yy + 56f,
                    w * .92f, yy - (float)Math.cos(phase) * 24f);
            c.drawPath(path, ribbon);
        }

        p.setColor(white ? 0xCCFFF0C4 : 0xD8B87AFF);
        for (int i = 0; i < px.length; i++) {
            float y = py[i] - (t * pv[i]) % 1.1f;
            if (y < -0.05f) y += 1.1f;
            float x = px[i] + (float)Math.sin(t * .8f + i) * .012f;
            p.setAlpha(70 + (i * 17) % 170);
            c.drawCircle(x * w, y * h, ps[i], p);
        }
        p.setAlpha(255);
    }
}

File: WuchangLiveWallpaperKotlin/app/src/main/java/com/halot455/wuchangwallpaper/HeroRenderer.kt
package com.halot455.wuchangwallpaper

import android.content.Context
import android.graphics.*
import android.util.Base64
import java.io.ByteArrayOutputStream
import kotlin.math.*

class HeroRenderer(context: Context) {
    private val appContext = context.applicationContext
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG)
    private val effectPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val cache = HashMap<String, Bitmap>()
    private val particles = List(44) { i ->
        Particle(
            x = ((i * 37) % 100) / 100f,
            y = ((i * 53) % 100) / 100f,
            size = 1.2f + (i % 5) * 0.7f,
            speed = 0.016f + (i % 7) * 0.004f,
            phase = i * 0.71f
        )
    }

    private data class Particle(
        val x: Float,
        val y: Float,
        val size: Float,
        val speed: Float,
        val phase: Float
    )

    private fun loadAsset(key: String): Bitmap? {
        cache[key]?.let { return it }
        return try {
            appContext.assets.open("$key.b64").use { input ->
                val out = ByteArrayOutputStream()
                val buffer = ByteArray(8192)
                while (true) {
                    val n = input.read(buffer)
                    if (n <= 0) break
                    out.write(buffer, 0, n)
                }
                val encoded = out.toString(Charsets.UTF_8.name())
                    .replace("\n", "")
                    .replace("\r", "")
                    .trim()
                val bytes = Base64.decode(encoded, Base64.DEFAULT)
                BitmapFactory.decodeByteArray(bytes, 0, bytes.size)?.also { cache[key] = it }
            }
        } catch (_: Throwable) {
            null
        }
    }

    fun draw(
        canvas: Canvas,
        width: Int,
        height: Int,
        angle: Float,
        nowMs: Long,
        character: String,
        pageOffset: Float
    ) {
        if (width <= 0 || height <= 0) return

        val isWhite = character != "black"
        val t = nowMs / 1000f
        val turn = (angle / 34f).coerceIn(-1f, 1f)

        paint.shader = LinearGradient(
            0f, 0f, 0f, height.toFloat(),
            if (isWhite) Color.rgb(16, 24, 42) else Color.rgb(18, 13, 29),
            Color.rgb(4, 6, 12),
            Shader.TileMode.CLAMP
        )
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
        paint.shader = null

        drawMoon(canvas, width, height, turn, pageOffset)
        drawFloorGlow(canvas, width, height, isWhite)
        drawCharacter(canvas, width, height, angle, turn, t, character)
        drawEnergy(canvas, width, height, isWhite, t, turn)
        drawParticles(canvas, width, height, isWhite, t)
    }

    private fun drawMoon(canvas: Canvas, width: Int, height: Int, turn: Float, pageOffset: Float) {
        val r = min(width, height) * 0.135f
        val cx = width * (0.77f - pageOffset * 0.04f - turn * 0.012f)
        val cy = height * 0.15f
        effectPaint.shader = RadialGradient(
            cx, cy, r * 1.55f,
            intArrayOf(0xEEFFFFFF.toInt(), 0x99DCE7FF.toInt(), 0x002E3C65),
            null,
            Shader.TileMode.CLAMP
        )
        canvas.drawCircle(cx, cy, r * 1.55f, effectPaint)
        effectPaint.shader = null
    }

    private fun drawFloorGlow(canvas: Canvas, width: Int, height: Int, white: Boolean) {
        effectPaint.shader = RadialGradient(
            width * 0.5f,
            height * 0.84f,
            width * 0.40f,
            intArrayOf(if (white) 0x3A87A7FF else 0x465E2AB6, 0x00000000),
            null,
            Shader.TileMode.CLAMP
        )
        canvas.drawOval(
            width * 0.08f, height * 0.72f,
            width * 0.92f, height * 0.96f,
            effectPaint
        )
        effectPaint.shader = null
    }

    private fun drawCharacter(
        canvas: Canvas,
        width: Int,
        height: Int,
        angle: Float,
        turn: Float,
        t: Float,
        character: String
    ) {
        val bitmap = loadAsset("${character}_front") ?: return

        val targetH = height * 0.855f
        val baseScale = targetH / bitmap.height
        val widthCompress = 1f - abs(turn) * 0.085f
        val targetW = bitmap.width * baseScale * widthCompress
        val left = (width - targetW) * 0.5f + turn * width * 0.018f
        val top = height * 0.078f + sin(t * 1.22f) * height * 0.0025f
        val src = Rect(0, 0, bitmap.width, bitmap.height)
        val dst = RectF(left, top, left + targetW, top + targetH)

        val breathe = 1f + sin(t * 1.55f) * 0.004f
        canvas.save()
        canvas.translate(turn * width * 0.006f, 0f)
        canvas.skew(-turn * 0.017f, 0f)
        canvas.scale(1f, breathe, width * 0.5f, top + targetH * 0.55f)
        paint.alpha = 255
        canvas.drawBitmap(bitmap, src, dst, paint)
        canvas.restore()

        val splitY = top + targetH * 0.43f
        val skirtSway = sin(t * 1.05f + angle * 0.018f)
        canvas.save()
        canvas.clipRect(0f, splitY, width.toFloat(), height.toFloat())
        canvas.translate((skirtSway + turn * 0.28f) * width * 0.0085f, 0f)
        canvas.skew(skirtSway * 0.009f - turn * 0.005f, 0f)
        paint.alpha = 78
        canvas.drawBitmap(bitmap, src, dst, paint)
        paint.alpha = 255
        canvas.restore()

        canvas.save()
        canvas.clipRect(0f, top, width.toFloat(), top + targetH * 0.36f)
        canvas.translate(sin(t * 1.38f) * width * 0.003f - turn * width * 0.002f, 0f)
        paint.alpha = 62
        canvas.drawBitmap(bitmap, src, dst, paint)
        paint.alpha = 255
        canvas.restore()
    }

    private fun drawEnergy(
        canvas: Canvas,
        width: Int,
        height: Int,
        white: Boolean,
        t: Float,
        turn: Float
    ) {
        effectPaint.style = Paint.Style.STROKE
        effectPaint.strokeWidth = max(2f, width * 0.0024f)
        effectPaint.color = if (white) 0x88EAF1FF.toInt() else 0xAA9D5BFF.toInt()

        repeat(3) { k ->
            val phase = t * (0.42f + k * 0.05f) + k * 1.7f
            val yy = height * (0.56f + k * 0.085f)
            val path = Path().apply {
                moveTo(width * 0.08f, yy + sin(phase) * 18f)
                cubicTo(
                    width * 0.27f, yy - 48f + turn * 8f,
                    width * 0.64f, yy + 54f - turn * 8f,
                    width * 0.92f, yy - cos(phase) * 24f
                )
            }
            canvas.drawPath(path, effectPaint)
        }
        effectPaint.style = Paint.Style.FILL
    }

    private fun drawParticles(canvas: Canvas, width: Int, height: Int, white: Boolean, t: Float) {
        paint.color = if (white) 0xCCFFF0C4.toInt() else 0xD8B87AFF.toInt()
        particles.forEachIndexed { i, part ->
            var y = part.y - (t * part.speed) % 1.12f
            if (y < -0.05f) y += 1.12f
            val x = part.x + sin(t * 0.8f + part.phase) * 0.012f
            paint.alpha = 70 + (i * 19) % 170
            canvas.drawCircle(x * width, y * height, part.size, paint)
        }
        paint.alpha = 255
    }
}
